Provider fails PSS signature with piv_II token

**Describe the bug**
```
openssl dgst  -sign "pkcs11:id=%02;type=private" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -out data_to_sign.bin.sig4 data_to_sign.bin
Enter pass phrase for PKCS#11 Token (Slot 1 - Yubico YubiKey OTP+FIDO+CCID):
$ !2965
openssl dgst  -verify "pkcs11:id=%02;type=public" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -signature data_to_sign.bin.sig4 data_to_sign.bin
Verified OK
$
$ ykcs11-tool --verify -m SHA384-RSA-PKCS-PSS -d 02 -y pubkey --signature-file data_to_sign.bin.sig4 -i data_to_sign.bin
Using slot 1 with a present token (0x1)
Using signature algorithm SHA384-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA384, mgf=MGF1-SHA384, salt_len=48 B
Invalid signature
$ pkcs11-tool --verify -m SHA384-RSA-PKCS-PSS -d 02 -y pubkey --signature-file data_to_sign.bin.sig4 -i data_to_sign.bin
Using slot 1 with a present token (0x4)
Using signature algorithm SHA384-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA384, mgf=MGF1-SHA384, salt_len=48 B
Invalid signature
$
```

**To Reproduce**
See above.

**Expected behavior**
```
$ openssl dgst -engine pkcs11 -keyform engine -sign "pkcs11:manufacturer=piv_II;id=%02;type=private" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -out data_to_sign.bin.sig4 data_to_sign.bin
Engine "pkcs11" set.
Enter PKCS#11 token PIN for Blumenthal, Uri (UR20980):
$
$ pkcs11-tool --verify -m SHA384-RSA-PKCS-PSS -d 02 -y pubkey --signature-file data_to_sign.bin.sig4 -i data_to_sign.bin
Using slot 1 with a present token (0x4)
Using signature algorithm SHA384-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA384, mgf=MGF1-SHA384, salt_len=48 B
Signature is valid
$ 
```

**Operating environment (please complete the following information):**
 - OS: MacOS (probably irrelevant)
 - Version: 13.2 (probably irrelevant)

**Token and application used (please complete the following information):**
 - Device: Yubikey-4 PIV applet. Probably, the only thing that matters is that it's (a) physical, and (b) smartdard (PIV).
 - PKCS11 Driver version: current master
 - Application: OpenSSL CLI
 - Version: 3.0.8

> > Note that only **some of them** are marked as `hw, <whatever>`. E.g., `ECDSA, hw, sign, verify`, but `ECDSA-SHA1, sign, verify`. I think it means - for the mechanisms that do **not** include `hw`, the token _assists_ software by performing the _actual_ asymmetric key operation (signature, decryption), but does **not** do the hashing and such.
> 
> I think that's opensc-pkcs11 doing the emulation, **and pkcs11-tool has some smarts to know that**. From the pkcs#11 consumer pov it doesn't matter what does what as long as the operation is performed according to specification.
> 
> pkcs11-provider can also emulate digest+sign opearations when the token supports only raw signing

I do not know about `pkcs11-tool` whether it knows when/what to emulate - IMHO `libp11 (aka "the engine") does know that.

> From the pkcs#11 consumer pov it doesn't matter what does what as long as the operation is performed according to specification.

Yes, but... A smartcard (usually) insists on the calling software (e.g., OpenSSL or provider or engine) to do the "mode" (PSS, OAEP, input hashing, etc. - aka, the "emulation") and only does the actual "raw" operations locally on the card. An HSM, on the other hand, **refuses** to do a "raw" operation (for security reasons - I discussed this with manufacturers) and does everything including input hashing on-board.

> pkcs11-provider can also emulate digest+sign opearations when the token supports only raw signing

Do you mean `pkcs11-provider` can do it right now, or that you can add this? If latter - could you please add this capability?

> Btw, RSA-PSS may not be fully functional yet with pkcs11-provider, I guess it is time to put in the effort ...

Yes please!

> But it would be nice to have traces only with a simpler PKCS1.5 signature method.

Understood, will do. (But if you got the engine running - I gave you the URL to the one I'm using - you can do it locally.)

> Interestingly enough, I can't create signatures

I concur:
```
$ openssl pkeyutl -sign -inkey "pkcs11:manufacturer=piv_II;id=%02;type=private" -pkeyopt rsa_padding_mode:pss -pkeyopt digest:sha384 -pkeyopt rsa_pss_saltlen:-1 -out /tmp/t.c.sha384.sig -in /tmp/t.c.sha384
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
Public Key operation error
40F6064DF87F0000:error:42000007:pkcs11:p11prov_Sign:Invalid or improper arguments were provided to the invoked function:./interface.gen.c:547:Error returned by C_Sign
$ openssl pkeyutl -sign -inkey "pkcs11:id=%02;type=private" -pkeyopt rsa_padding_mode:pss -pkeyopt digest:sha384 -pkeyopt rsa_pss_saltlen:-1 -out /tmp/t.c.sha384.sig -in /tmp/t.c.sha384
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
Public Key operation error
40F6064DF87F0000:error:42000007:pkcs11:p11prov_Sign:Invalid or improper arguments were provided to the invoked function:./interface.gen.c:547:Error returned by C_Sign
$
$ $ openssl pkeyutl -engine pkcs11 -keyform engine -sign -inkey "pkcs11:manufacturer=piv_II;id=%02;type=private" -pkeyopt rsa_padding_mode:pss -pkeyopt digest:sha384 -pkeyopt rsa_pss_saltlen:-1 -out /tmp/t.c.sha384.sig -in /tmp/t.c.sha384
Engine "pkcs11" set.
Enter PKCS#11 token PIN for XXXXXXXXX:
$ ykcs11-tool --verify -m SHA384-RSA-PKCS-PSS -d 02 -y pubkey --signature-file /tmp/t.c.sha384.sig -i ~/src/t.c
Using slot 1 with a present token (0x1)
Using signature algorithm SHA384-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA384, mgf=MGF1-SHA384, salt_len=48 B
Signature is valid
$ 
```

On the other hand (sign using RSA-PKCS1):
```
$ ll /tmp/t.c.sha384
-rw-r--r--  1 ur20980  wheel  48 Feb 11 22:51 /tmp/t.c.sha384
$
$ openssl pkeyutl -sign -inkey "pkcs11:id=%02;type=private" -pkeyopt rsa_padding_mode:pkcs1 -out /tmp/t.c.sha384.sig -in /tmp/t.c.sha384
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
$
$ pkcs11-tool --verify -m RSA-PKCS -d 02 -y pubkey --signature-file /tmp/t.c.sha384.sig -i /tmp/t.c.sha384
Using slot 1 with a present token (0x4)
Using signature algorithm RSA-PKCS
Signature is valid
$ 
```

P.S. I think it would help us all and speed up the testing greatly if you could get the engine installed and running on your machine. Having both in the `openssl.cnf` doesn't interfere with the provider testing.

The command I use is this:
```
$ openssl dgst -sha256 -sign "pkcs11:id=%02;type=private"  -out t.sig t
Enter pass phrase for PKCS#11 Token (Slot 0 - Yubico YubiKey OTP+FIDO+CCID 00 00):
Error signing data
0002BD1D5A7F0000:error:40800054:pkcs11:p11prov_GetOperationState:reason(84):interface.gen.c:223:Error returned by C_GetOperationState
0002BD1D5A7F0000:error:40800020:pkcs11:p11prov_SignFinal:The plaintext input data to a cryptographic operation is invalid:interface.gen.c:577:Error returned by C_SignFinal
```

Here is the debug file:
[opensc-debug.txt](https://github.com/latchset/pkcs11-provider/files/10723834/opensc-debug.txt)


If I remove use_file_caching directive I get this:
```
$ openssl dgst -sha256 -sign "pkcs11:id=%02;type=private"  -out t.sig t
Enter pass phrase for PKCS#11 Token (Slot 0 - Yubico YubiKey OTP+FIDO+CCID 00 00):
Could not read private key from pkcs11:id=%02;type=private
```

--- a/src/signature.c
+++ b/src/signature.c
@@ -430,52 +430,79 @@ static CK_RV p11prov_sig_pss_restrictions(P11PROV_SIG_CTX *sigctx,
     return CKR_OK;
 }
 
-static int p11prov_sig_set_mechanism(void *ctx, bool digest_sign,
+/* fixates pss_params based on defaults if values are not set */
+static CK_RV pss_defaults(P11PROV_SIG_CTX *sigctx, CK_MECHANISM *mechanism,
+                          bool set_mech)
+{
+    const P11PROV_MECH *mech;
+    CK_RV ret;
+
+    ret = p11prov_mech_by_mechanism(sigctx->digest, &mech);
+    if (ret != CKR_OK) {
+        return ret;
+    }
+    sigctx->pss_params.hashAlg = mech->digest;
+    if (sigctx->pss_params.mgf == 0) {
+        sigctx->pss_params.mgf = mech->mgf;
+    }
+    if (sigctx->pss_params.sLen == 0) {
+        /* default to digest size if not set */
+        size_t size;
+        ret = p11prov_digest_get_digest_size(mech->digest, &size);
+        if (ret != CKR_OK) {
+            return ret;
+        }
+        sigctx->pss_params.sLen = size;
+    }
+
+    mechanism->pParameter = &sigctx->pss_params;
+    mechanism->ulParameterLen = sizeof(sigctx->pss_params);
+
+    if (set_mech) {
+        mechanism->mechanism = mech->pkcs_pss;
+    }
+
+    return CKR_OK;
+}
+
+static int p11prov_sig_set_mechanism(P11PROV_SIG_CTX *sigctx, bool digest_sign,
                                      CK_MECHANISM *mechanism)
 {
-    P11PROV_SIG_CTX *sigctx = (P11PROV_SIG_CTX *)ctx;
-    const P11PROV_MECH *mech;
     int result = CKR_DATA_INVALID;
 
     mechanism->mechanism = sigctx->mechtype;
     mechanism->pParameter = NULL;
     mechanism->ulParameterLen = 0;
 
-    if (sigctx->mechtype == CKM_RSA_PKCS_PSS) {
-        mechanism->pParameter = &sigctx->pss_params;
-        mechanism->ulParameterLen = sizeof(sigctx->pss_params);
-        if (sigctx->digest) {
-            sigctx->pss_params.hashAlg = sigctx->digest;
-        }
-    }
-
-    if (!digest_sign) {
-        return CKR_OK;
-    }
-
     switch (sigctx->mechtype) {
     case CKM_RSA_PKCS:
-        result = p11prov_mech_by_mechanism(sigctx->digest, &mech);
-        if (result == CKR_OK) {
-            mechanism->mechanism = mech->pkcs_mech;
+        if (digest_sign) {
+            const P11PROV_MECH *mech;
+            result = p11prov_mech_by_mechanism(sigctx->digest, &mech);
+            if (result == CKR_OK) {
+                mechanism->mechanism = mech->pkcs_mech;
+            }
+        } else {
+            result = CKR_OK;
         }
         break;
     case CKM_RSA_X_509:
         break;
     case CKM_RSA_PKCS_PSS:
-        mechanism->pParameter = &sigctx->pss_params;
-        mechanism->ulParameterLen = sizeof(sigctx->pss_params);
-
-        result = p11prov_mech_by_mechanism(sigctx->digest, &mech);
-        if (result == CKR_OK) {
-            mechanism->mechanism = mech->pkcs_pss;
-            result = p11prov_sig_pss_restrictions(ctx, mechanism);
+        result = pss_defaults(sigctx, mechanism, digest_sign);
+        if (result == CKR_OK && digest_sign) {
+            result = p11prov_sig_pss_restrictions(sigctx, mechanism);
         }
         break;
     case CKM_ECDSA:
-        result = p11prov_mech_by_mechanism(sigctx->digest, &mech);
-        if (result == CKR_OK) {
-            mechanism->mechanism = mech->ecdsa_mech;
+        if (digest_sign) {
+            const P11PROV_MECH *mech;
+            result = p11prov_mech_by_mechanism(sigctx->digest, &mech);
+            if (result == CKR_OK) {
+                mechanism->mechanism = mech->ecdsa_mech;
+            }
+        } else {
+            result = CKR_OK;
         }
         break;
     }
@@ -534,6 +561,48 @@ static int p11prov_rsasig_set_pss_saltlen_from_digest(void *ctx)
     }
 
     sigctx->pss_params.sLen = digest_size;
+    return RET_OSSL_OK;
+}
+
+static int p11prov_rsasig_set_pss_saltlen_max(void *ctx, bool max_to_digest)
+{
+    P11PROV_SIG_CTX *sigctx = (P11PROV_SIG_CTX *)ctx;
+    size_t digest_size;
+    CK_ULONG key_size;
+    CK_ULONG key_bit_size;
+    CK_RV rv;
+
+    if (sigctx->digest == 0) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_NOT_SUPPORTED,
+                       "Can only be set if Digest was set first.");
+        return RET_OSSL_ERR;
+    }
+
+    rv = p11prov_digest_get_digest_size(sigctx->digest, &digest_size);
+    if (rv != CKR_OK) {
+        P11PROV_raise(sigctx->provctx, rv, "Unavailable digest");
+        return RET_OSSL_ERR;
+    }
+
+    key_size = p11prov_obj_get_key_size(sigctx->key);
+    if (key_size == CK_UNAVAILABLE_INFORMATION) {
+        P11PROV_raise(sigctx->provctx, rv, "Unavailable key");
+        return RET_OSSL_ERR;
+    }
+    key_bit_size = p11prov_obj_get_key_bit_size(sigctx->key);
+    if (key_bit_size == CK_UNAVAILABLE_INFORMATION) {
+        P11PROV_raise(sigctx->provctx, rv, "Unavailable key");
+        return RET_OSSL_ERR;
+    }
+
+    /* from openssl */
+    sigctx->pss_params.sLen = key_size - digest_size - 2;
+    if ((key_bit_size & 0x07) == 1) {
+        sigctx->pss_params.sLen -= 1;
+    }
+    if (max_to_digest && sigctx->pss_params.sLen > digest_size) {
+        sigctx->pss_params.sLen = digest_size;
+    }
     return RET_OSSL_OK;
 }
 
@@ -980,11 +1049,6 @@ static void *p11prov_rsasig_newctx(void *provctx, const char *properties)
         return NULL;
     }
 
-    /* default PSS Params */
-    sigctx->pss_params.hashAlg = CKM_SHA256;
-    sigctx->pss_params.mgf = CKG_MGF1_SHA256;
-    sigctx->pss_params.sLen = 32;
-
     return sigctx;
 }
 
@@ -1245,7 +1309,17 @@ static int p11prov_rsasig_get_ctx_params(void *ctx, OSSL_PARAM *params)
 
     p = OSSL_PARAM_locate(params, OSSL_SIGNATURE_PARAM_MGF1_DIGEST);
     if (p) {
-        const char *digest = p11prov_sig_mgf_name(sigctx->pss_params.mgf);
+        const char *digest = NULL;
+
+        if (sigctx->pss_params.mgf != 0) {
+            digest = p11prov_sig_mgf_name(sigctx->pss_params.mgf);
+        } else {
+            const P11PROV_MECH *pssmech;
+            CK_RV rv = p11prov_mech_by_mechanism(sigctx->mechtype, &pssmech);
+            if (rv == CKR_OK) {
+                p11prov_digest_get_name(pssmech->digest, &digest);
+            }
+        }
         if (!digest) {
             return RET_OSSL_ERR;
         }
@@ -1256,6 +1330,32 @@ static int p11prov_rsasig_get_ctx_params(void *ctx, OSSL_PARAM *params)
     }
 
     return RET_OSSL_OK;
+}
+
+/* only available in recent OpenSSL 3.0.x headers */
+#ifndef RSA_PSS_SALTLEN_AUTO_DIGEST_MAX
+#define RSA_PSS_SALTLEN_AUTO_DIGEST_MAX -4
+#endif
+#ifndef OSSL_PKEY_RSA_PSS_SALT_LEN_AUTO_DIGEST_MAX
+#define OSSL_PKEY_RSA_PSS_SALT_LEN_AUTO_DIGEST_MAX "auto-digestmax"
+#endif
+
+static int rsasig_set_saltlen(P11PROV_SIG_CTX *sigctx, int saltlen)
+{
+    if (saltlen >= 0) {
+        sigctx->pss_params.sLen = saltlen;
+        return RET_OSSL_OK;
+    }
+    if (saltlen == RSA_PSS_SALTLEN_DIGEST) {
+        return p11prov_rsasig_set_pss_saltlen_from_digest(sigctx);
+    }
+    if (saltlen == RSA_PSS_SALTLEN_AUTO || saltlen == RSA_PSS_SALTLEN_MAX) {
+        return p11prov_rsasig_set_pss_saltlen_max(sigctx, false);
+    }
+    if (saltlen == RSA_PSS_SALTLEN_AUTO_DIGEST_MAX) {
+        return p11prov_rsasig_set_pss_saltlen_max(sigctx, true);
+    }
+    return RET_OSSL_ERR;
 }
 
 static int p11prov_rsasig_set_ctx_params(void *ctx, const OSSL_PARAM params[])
@@ -1347,32 +1447,28 @@ static int p11prov_rsasig_set_ctx_params(void *ctx, const OSSL_PARAM params[])
             }
             P11PROV_debug("Set OSSL_SIGNATURE_PARAM_PSS_SALTLEN to %d",
                           saltlen);
-            sigctx->pss_params.sLen = saltlen;
         } else if (p->data_type == OSSL_PARAM_UTF8_STRING) {
             P11PROV_debug("Set OSSL_SIGNATURE_PARAM_PSS_SALTLEN to %s",
                           p->data ? (const char *)p->data : "<NULL>");
             if (strcmp(p->data, OSSL_PKEY_RSA_PSS_SALT_LEN_DIGEST) == 0) {
-                ret = p11prov_rsasig_set_pss_saltlen_from_digest(sigctx);
-                if (ret != RET_OSSL_OK) {
-                    return ret;
-                }
+                saltlen = RSA_PSS_SALTLEN_DIGEST;
             } else if (strcmp(p->data, OSSL_PKEY_RSA_PSS_SALT_LEN_MAX) == 0) {
-                ERR_raise_data(ERR_LIB_PROV, PROV_R_NOT_SUPPORTED,
-                               "saltlen=max is unsupported.");
-                return RET_OSSL_ERR;
+                saltlen = RSA_PSS_SALTLEN_MAX;
             } else if (strcmp(p->data, OSSL_PKEY_RSA_PSS_SALT_LEN_AUTO) == 0) {
-                ERR_raise_data(ERR_LIB_PROV, PROV_R_NOT_SUPPORTED,
-                               "saltlen=auto is unsupported.");
-                return RET_OSSL_ERR;
+                saltlen = RSA_PSS_SALTLEN_AUTO;
+            } else if (strcmp(p->data,
+                              OSSL_PKEY_RSA_PSS_SALT_LEN_AUTO_DIGEST_MAX)
+                       == 0) {
+                saltlen = RSA_PSS_SALTLEN_AUTO_DIGEST_MAX;
             } else {
                 saltlen = atoi(p->data);
-                if (saltlen == 0) {
-                    return RET_OSSL_ERR;
-                }
-                sigctx->pss_params.sLen = saltlen;
             }
         } else {
             return RET_OSSL_ERR;
+        }
+        ret = rsasig_set_saltlen(sigctx, saltlen);
+        if (ret != RET_OSSL_OK) {
+            return ret;
         }
     }
 

